src/irmin-pack/irmin_pack.ml

```diff
@@ -557,8 +573,60 @@ struct
         in
         Contents.CA.sync ~on_generation_change (contents_t t)
 
-      (** Storea share instances so one clear is enough. *)
+      (** Stores share instances so one clear is enough. *)
       let clear t = Contents.CA.clear (contents_t t)
+
+      (** Migrate data from the IO [src] (with [name] in path [root_old]) into
+          the temporary dir [root_tmp], then swap in the replaced version. *)
+      let migrate_io_to_v2 ~progress src =
+        IO.migrate ~progress src `V2 |> function
+        | Ok () -> IO.close src
+        | Error (`Msg s) -> invalid_arg s
+
+      let migrate config =
+        let* () =
+          if readonly config then Lwt.fail RO_Not_Allowed else Lwt.return_unit
+        in
+        let config = Irmin.Private.Conf.add config readonly_key true in
+        Log.info (fun l -> l "[%s] migrate" (root config));
+        let* t = unsafe_v config in
+        match Contents.CA.version (contents_t t) with
+        | `V2 as v ->
+            Log.app (fun l ->
+                l "Store at %s is already in current version (%a)" (root config)
+                  pp_version v);
+            close t
+        | `V1 ->
+            let* () = close t in
+            let root_old = root config in
+            let ios =
+              [ "store.pack"; "store.branches"; "store.dict" ]
+              |> List.map (fun name ->
+                     IO.v ~version:`V1 ~fresh:false ~readonly:true
+                       (root_old // name))
+            in
+            let total =
+              ios |> List.map IO.offset |> List.fold_left Int64.add 0L
+            in
+            let bar, progress =
+              Utils.Progress.counter ~total ~sampling_interval:10
+                ~message:"Migrating store" ~pp_count:Utils.pp_bytes ()
+            in
+            ios |> List.iter (migrate_io_to_v2 ~progress);
+            Utils.Progress.finalise bar;
+            List.iter
+              (fun i ->
+                i ~readonly:true root_old;
```

Comment:
Given how it seems to be used here, maybe `invalidate` should just take the path as argument and invalidate both read-write and read-only open instances? Is there a case where we want to invalidate only one kind of instances?

Reply:
Indeed: we only ever want to invalidate both instances. My thinking here was that the less the caches know about the semantics of the `(string * bool)` key the better (and the easier it will be to simplify them later), but it's probably fine to just inline this logic here.